tangelo/toolboxes/circuits/lcu.py

```diff
@@ -0,0 +1,310 @@
+                if abs(phasep) > 1.e-12 or not np.isclose(abs(phasep), np.pi*2):
+                    gate_list.append(Gate("CPHASE", target=0, control=k_control_qubits, parameter=phasep))
+                gate_list += [Gate("C"+op, target=index, control=k_control_qubits) for index, op in term]
+            gate_list += [Gate("X", q_start + q) for q, i in enumerate(state_binstr) if i == "0"]
+
+    # Add -I term to ensure total sum equals 2 for oblivious amplitude amplification
+    k_control_qubits = [k_start+ki for ki in range(kmax)] + [n_qubits_sv + kmax + kmax*n_qubits_u] + control_list
+    x_ladder = [Gate("X", k_start + q) for q in range(kmax)]
+    gate_list += x_ladder
+    gate_list.append(Gate("CRZ", target=0, control=k_control_qubits, parameter=2*np.pi))
+    gate_list += x_ladder
+
+    return Circuit(gate_list)
+
+
+def sign_flip(qubit_list: List[QubitOperator], control: Union[int, List[int]] = None) -> Circuit:
+    """Generate Circuit corresponding to the sign flip of the |0>^n vector for the given qubit_list
+    Args:
+        qubit_list (list): The list of n qubits for which the 2*|0>^n-I operation is generated
+        control (int or list[int]): Control qubit or list of control qubits.
+    Returns:
+        Circuit: The circuit that generates the sign flip on |0>^n
+    """
+    if control is not None:
+        fcontrol_list = control if isinstance(control, list) else [control]
+    else:
+        fcontrol_list = []
+    gate_list = []
+
+    x_ladder = [Gate("X", q) for q in qubit_list]
+    fcontrol_list += [q for q in qubit_list[:-1]]
+
+    gate_list += x_ladder
+    gate_list.append(Gate('CZ', target=qubit_list[-1], control=fcontrol_list))
+    gate_list += x_ladder
+    return Circuit(gate_list)
+
+
+def get_lcu_circuit(qu_op: QubitOperator, control: Union[int, List[int]] = None) -> Circuit:
+    """Apply qu_op using the linear combination of unitaries algorithm
+    1-norm of coefficients must be less than 2. The unitarity of qu_op is not checked by the algorithm
+    Args:
+        qu_op (QubitOperator): The qu_op to apply. Must be nearly unitary for algorithm to succeed with high probability.
+        control (int or list[int]): Control qubit(s)
+    Returns:
+        Circuit: The circuit that implements the linear combination of unitaries for the qu_op
+    """
+
+    if control is not None:
+        control_list = control if isinstance(control, list) else [control]
+    else:
+        control_list = []
+
+    unitaries = list()
+    vector = list()
+    max_qu_op = count_qubits(qu_op)
+    for term, coeff in qu_op.terms.items():
+        acoeff = np.abs(coeff)
+        vector += [acoeff]
+        unitaries += [QubitOperator(term, coeff / acoeff)]
+    vsum = sum(vector)
+
+    # Check that 1-norm is less than 2 and add and subtract terms proportional to the identity to obtain 1-norm equal to 2
+    if vsum > 2:
```

Comment:
I think you may want to account for a machine precision epsilon here (e.g what if the sum is 2.00000000001?)

`if vsum > 2 + 1e-k:` (I let you decide what is a reasonable value for `k`)